### PR TITLE
FEATURE: Dynamic cached segments

### PR DESCRIPTION
**Pull request by kitsunet**
This change adds a new type of cache mode, the so called ``dynamic``.
It will always trigger a new evaluation at runtime but the results are
cached again. The configuration for a ``dynamic`` segment has a
new property ``entryDiscriminator``. This defines how the cached entries
for this segment are discerned from one another at runtime.
Usually you will use an EEL expression, but this can only use global
context variables like the request.

An example configuration for segment that is separately cached depending
on a GET argument ``someArgument`` could look like this::

    @cache {
    	mode = 'dynamic'
    	entryIdentifier {
    		node = ${node}
    	}
    	entryDiscriminator = ${request.httpRequest.arguments.someArgument}
    	context {
    		1 = 'node'
    	}
    }

### CODE REVIEW

**Comment by kitsunet** [comment]
This is my WIP for cache parts that depend on some specific information. @kdambekalns we spoke about this. It is generally working but could need tests and a general discussion about the feature

**Comment by skurfuerst** [comment]
@kitsunet I'd be up for discussing this; I do not see yet where this is needed -- so I guess I need some more background :-)

All the best, Sebastian

**Comment by kuborgh-mspindelhirn** [comment]
If I understand correctly this would be a useful cache setting for the `root` TS object if you implement caches on page objects that depend on request values.

**Comment by kitsunet** [comment]
Specifically the root should not be needing it. But only the specific content element that depends on a request value (like a plugin or list with pagination).

**Comment by johannessteu** [comment]
We'll @skurfuerst you would need for something like this https://github.com/Flowpack/Flowpack.ElasticSearch.ContentRepositoryAdaptor#product-filter

This way you could define 
```
@cache {
    mode = 'runtimeCache'
    entryIdentifier {
        node = ${node}
        searchFilter = ${searchFilter}
    }
    entryDiscriminator {
        1 = ${request.httpRequest.arguments.color}
        2 = ${request.httpRequest.arguments.size}
    }
    context {
        1 = 'node'
    }
}
```

and your search results would be cached. So adding multiple values to `entryDiscriminator` would be great @kitsunet . I would heavily vote for this feature and i think there are a lot use cases!

**Comment by kitsunet** [comment]
great, yes, makes sense, shouldn't be a big problem to add multiple discriminators.

**Comment by johannessteu** [comment]
Just thougt that it would also be posible to create some fancy eel query to get mulitple in?! But yeah, an array would be easier

**Comment by kdambekalns** [comment]
This lacks **documentation**.

And I'd love to have an explanation on how this is different from just having the values that make up the `entryDiscriminator` set as `nodeIdentifier` parts.

**Comment by kdambekalns** [comment]
My last comment still holds, but I merged master and did some tiny adjustments. @kitsunet Why is this marked WIP?

**Comment by kitsunet** [comment]
Removed WIP label. Will add documentation though. The difference is that you do not need to add the discriminator on every level up the TS chain as that can lead to a lot of useless cache entries. A runtime cached segment is fully standalone and doesn't depend on the parents cache configuration (important for plugins).

**Comment by kdambekalns** [comment]
Ok, got it. Looking forward to the docs, then… :)

**Comment by kitsunet** [comment]
Ok, went over the naming again and documented...

File: TYPO3.Neos/Documentation/CreatingASite/ContentCache.rst [hunk 1]
```diff
@@ -76,24 +76,28 @@ The given configuration will cache the entire page content with a unique identif
...
+  ``'dynamic'`` or ``'uncached'``.
   Only simple string values are supported for this property.
 
   It defaults to mode ``embed`` which will not create a new cache entry but store the content into the next outer cached
   entry. With mode ``cached`` a separate cache entry will be created for the path. Mode ``uncached`` can be used to
-  always evaluate a path even if is contained inside a cached path. The ``context`` property should be set to configure
-  the TypoScript context variables that will be available when evaluating the uncached path.
+  always evaluate a path even if is contained inside a cached path. The ``dynamic`` mode evalutes a so called
+  "discriminator" on every request and caches results differently depending on it's value. Dynamic cache mode is therefore
+  much faster then uncached but slightly slower then cached mode. It is useful in situations where arguments (eg. from the
```
Comment (robertlemke) [inline TYPO3.Neos/Documentation/CreatingASite/ContentCache.rst]: "then" -> "than"
uncached -> ``uncached``
cached -> ``cached``

File: TYPO3.Neos/Documentation/CreatingASite/ContentCache.rst [hunk 2]
```diff
@@ -135,6 +139,11 @@ In the ``@cache`` meta property the following subproperties are allowed:
 		}
 	}
 
+``entryDiscriminator``
+  Configure an expression that uniquely discriminates different entries of a ``dynamic`` cached area. The expression or TypoScript
+  object must evaluate to a string to be used as discriminator and should be different for every cache entry you want to create for
+  this ``dynamic`` cached area.
+
 .. _Cache Entry Tags:
```
Comment (robertlemke) [inline TYPO3.Neos/Documentation/CreatingASite/ContentCache.rst]: Not the fault of this patch, but I think to truly understand how the different caching options are used, we really need some short real world examples with working TypoScript snippets.

**Comment by robertlemke** [comment]
👍 for the concept (although I was reluctant first). I quickly read all the code and documentation and left a few notes regarding the docs. Didn't get around to test it yet.

**Comment by kitsunet** [comment]
Did some more polishing to the documentation.

**Pull request merged**